app/models/manageiq/providers/redhat/infra_manager/vm/reconfigure.rb

```diff
@@ -23,6 +23,29 @@ def max_memory_mb
     2.terabyte / 1.megabyte
   end
 
+  def available_vlans
+    vlans = []
+
+    switch_ids = HostSwitch.where(:host => host).pluck(:switch_id)
+    Lan.where(:switch_id => switch_ids).each do |lan|
+      vlans << lan.name
+    end
+
+    vlans.sort.concat(available_external_vlans)
+  end
+
+  def available_external_vlans
+    ext_vlans = []
+    ems = ext_management_system
+
+    ext_switch_ids = ems.external_distributed_virtual_switches.pluck(:id)
+    ems.external_distributed_virtual_lans.where(:switch_id => ext_switch_ids).each do |ext_lan|
+      ext_vlans << "#{ext_lan.name}/#{ext_lan.switch.name}"
+    end
```

Comment:
Same question here, `external_distributed_virtual_lans` is already `:through => :external_distributed_virtual_switches` so I think you're re-implementing the rails association here without the caching :)

Try just `ems.external_distributed_virtual_lans.map { |lan| "#{lan.name}/#{lan.switch.name}" }`

Reply:
I used the same style that was used in the manageiq-ui-classic. I was not sure if it was better to use rails relations involving more objects or use the pluck.

Reply:
@gekorob can you link to the code that you found in UI classic?  Curious on the history there

Reply:
@gekorob can you link to the code that you found in UI classic?  Curious on the history there

Reply:
@agrare yes
https://github.com/ManageIQ/manageiq-ui-classic/blob/e7e66e5014f25fc2c2e0ec9a1a115602dea93c57/app/controllers/mixins/actions/vm_actions/reconfigure.rb#L214-L226
I think the purpose was also to use filter access on Roles

Reply:
I tried to propose a simple way to delegate the list of vlans to the provider if it implements the method "available_vlans"
https://github.com/gekorob/manageiq-ui-classic/blob/2122a372aaa87ddca3abe5ea43b51cba53aa9518/app/controllers/mixins/actions/vm_actions/reconfigure.rb#L22

Maybe not the best thing, but similar what was done for adapters_name and currently implemented in the vmware provider I think
https://github.com/gekorob/manageiq-ui-classic/blob/2122a372aaa87ddca3abe5ea43b51cba53aa9518/app/controllers/mixins/actions/vm_actions/reconfigure.rb#L24